src/base/array_view.hpp

```diff
@@ -50,48 +48,43 @@ class ArrayView
   using const_iterator = const_pointer;
 
   ArrayView() = default;
-  constexpr ArrayView(const T* pData, const size_type size) noexcept
+  constexpr ArrayView(const_pointer pData, const size_type size) noexcept
     : mpData(pData)
     , mSize(size)
   {
   }
 
   // implicit on purpose
   template <std::size_t N>
-  constexpr ArrayView(const std::array<T, N>& array) noexcept // NOLINT
-    : mpData(array.data())
-    , mSize(static_cast<size_type>(N))
-  {
-  }
-
-  // implicit on purpose
-  template <std::size_t N>
-  constexpr ArrayView(const T (&array)[N]) noexcept // NOLINT
+  constexpr ArrayView(const value_type (&array)[N]) noexcept // NOLINT
     : mpData(array)
     , mSize(static_cast<size_type>(N))
   {
   }
 
   // implicit on purpose
-  ArrayView(const std::vector<T>& vec) noexcept // NOLINT
-    : mpData(vec.data())
-    , mSize(static_cast<size_type>(vec.size()))
+  template <typename Container>
+  constexpr ArrayView(const Container& c) noexcept // NOLINT
+    : mpData(c.data())
+    , mSize(static_cast<size_type>(c.size()))
   {
   }
 
-  const_iterator begin() const { return mpData; }
+  constexpr const_iterator begin() const noexcept { return mpData; }
```

Comment:
While we're at it, should we go all the way and make these also `[[nodiscard]]`? (And `front`, `back`, `data`, `operator[]`)?